contracts/multisig-prover/src/execute.rs

```diff
@@ -126,31 +128,116 @@ pub fn update_worker_set(deps: DepsMut, env: Env) -> Result<Response, ContractEr
+        env.block.height,
+    )?;
+
+    let cur_worker_set = CURRENT_WORKER_SET.may_load(deps.storage)?;
 
-    let cur_worker_set = CURRENT_WORKER_SET.load(deps.storage)?;
+    let key_id = new_worker_set.id();
 
-    if !should_update_worker_set(
-        &new_worker_set,
-        &cur_worker_set,
-        config.worker_set_diff_threshold as usize,
-    ) {
-        return Err(ContractError::WorkerSetUnchanged);
+    match cur_worker_set {
+        None => {
+            // if no worker set, just store it and return
+            CURRENT_WORKER_SET.save(deps.storage, &new_worker_set)?;
+
+            KEY_ID.save(deps.storage, &key_id)?;
+
+            let key_gen_msg = multisig::msg::ExecuteMsg::KeyGen {
+                key_id,
+                snapshot,
+                pub_keys_by_address: participants
+                    .into_iter()
+                    .map(|(participant, pub_key)| {
+                        (
+                            participant.address.to_string(),
+                            (KeyType::Ecdsa, pub_key.as_ref().into()),
+                        )
+                    })
+                    .collect(),
+            };
+
+            Ok(Response::new().add_message(wasm_execute(config.multisig, &key_gen_msg, vec![])?))
+        }
+        Some(cur_worker_set) => {
+            if !should_update_worker_set(
+                &new_worker_set,
+                &cur_worker_set,
+                config.worker_set_diff_threshold as usize,
+            ) {
+                return Err(ContractError::WorkerSetUnchanged);
+            }
+
+            NEXT_WORKER_SET.save(deps.storage, &(new_worker_set.clone(), snapshot))?;
+            let mut builder = CommandBatchBuilder::new(config.destination_chain_id);
+            builder.add_new_worker_set(new_worker_set)?;
+
+            let batch = builder.build()?;
+
+            COMMANDS_BATCH.save(deps.storage, &batch.id, &batch)?;
+            REPLY_BATCH.save(deps.storage, &batch.id)?;
+
+            let start_sig_msg = multisig::msg::ExecuteMsg::StartSigningSession {
+                key_id: cur_worker_set.id(), // TODO remove the key_id
+                msg: batch.msg_to_sign(),
+            };
+
+            Ok(Response::new().add_submessage(SubMsg::reply_on_success(
+                wasm_execute(config.multisig, &start_sig_msg, vec![])?,
+                START_MULTISIG_REPLY_ID,
+            )))
+        }
     }
+}
+
+pub fn confirm_worker_set(deps: DepsMut) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
 
-    NEXT_WORKER_SET.save(deps.storage, &new_worker_set)?;
-    let mut builder = CommandBatchBuilder::new(config.destination_chain_id);
-    builder.add_new_worker_set(new_worker_set)?;
+    let (worker_set, snapshot) = NEXT_WORKER_SET.load(deps.storage)?;
 
-    let batch = builder.build()?;
+    let query = voting_verifier::msg::QueryMsg::IsWorkerSetConfirmed {
+        new_operators: worker_set.clone().try_into()?,
```

Comment:
Here the operators are not nonced so a new worker set with the exact same operators as any previous one is immediately confirmed?

Reply:
Good catch. This is actually pretty tricky to deal with, because the nonce is not passed to the evm gateway.

There are a lot of ways to deal with this:

1. Make the voting verifier actually call `ConfirmWorkerSet` on the prover, after the worker set is confirmed. As far as I can tell, this is the closest to how we handle this in core. The main challenge here is this creates a circular dependency at instantiation time. The voting verifier would need to know the address of the prover, but the prover needs to know the address of the gateway, which needs to know the address of the voting verifier. If there was some sort of constant address deployer for cosmwasm contracts, this would be easier.
2. Prune old operator confirmations from the voting verifier. We really only need the most recent. The EVM gateway stores the last 16, so that would work as well. We could also still store the old ones, but only query the most recent (or last 16).

Either way, I do think we should tackle this in a separate PR.

Reply:
Well actually apparently the EVM gateway doesn't allow duplicate operators lol: https://github.com/axelarnetwork/axelar-cgp-solidity/blob/134d264144b5adc32c42b1389618e86c91fbdc75/contracts/auth/AxelarAuthWeighted.sol#L80

I think this doesn't really make sense, but given this is the case, we should just enforce the same thing in the prover.

Reply:
Yeah I saw that line as well and thought it was weird, but Milap mentioned it's because in consensus the public keys are not reused during key rotation: https://axelarworkspace.slack.com/archives/C02T1BNNS9Y/p1693326543812459. So that check makes sense for consensus, but not anymore for amplifier. So I don't think we should enforce the same thing.

Reply:
I think until we change the evm gateway, we actually have to check this in amplifier. Otherwise amplifier could start a rotation that will never be confirmed by the evm gateway, which will cause problems.

Reply:
Task to keep track of this: https://axelarnetwork.atlassian.net/browse/AXE-1869

Reply:
@cjcobb23 @eguajardo A lot of the solidity contracts were implemented with assumptions from the axelar-core's consensus code, which no longer holds true for the new cosmwasm implementation.